File: app/src/main/java/com/iamagamedev/mybrandnewgame/Animation.java
package com.iamagamedev.mybrandnewgame;

import android.graphics.Bitmap;
import android.graphics.Rect;

/**
 * Created by Михан on 25.05.2017.
 */

public class Animation {
    Bitmap bitmapSheet;
    private String bitmapName;
    private Rect soursRect;
    private int frameCount;
    private int currentFrame;
    private long frameTicker;
    private int framePeriod;
    private float frameWidth;
    private float frameHeight;
    private int pixelsPerMetre;

    public Animation(String bitmapName, float frameHeight,
                     float frameWidth, int animFps, int frameCount, int pixelsPerMetre) {
        this.currentFrame = 0;
        this.frameCount = frameCount;
        this.frameWidth = frameWidth * pixelsPerMetre;
        this.frameHeight = frameHeight * pixelsPerMetre;
        soursRect = new Rect(0, 0, (int) this.frameWidth, (int) this.frameHeight);
        framePeriod = 1000 / animFps;
        frameTicker = 0L;
        this.bitmapName = "" + bitmapName;
        this.pixelsPerMetre = pixelsPerMetre;
    }

    public Rect getCurrentFrame(long time, float xVelocity,
                                float yVelocity, boolean moves) {
        if (xVelocity != 0 || yVelocity != 0 || !moves) {

            if (time > frameTicker + framePeriod) {
                frameTicker = time;
                currentFrame++;
                if (currentFrame >= frameCount) {
                    currentFrame = 0;
                }
            }
        }
        this.soursRect.left = currentFrame * (int) frameWidth;
        this.soursRect.right = this.soursRect.left + (int) frameWidth;

        return soursRect;
    }
}

File: app/src/main/java/com/iamagamedev/mybrandnewgame/Animation.kt
package com.iamagamedev.mybrandnewgame

import android.graphics.Bitmap
import android.graphics.Rect

/**
 * Created by Михан on 25.05.2017.
 */
class Animation(bitmapName: String, frameHeight: Float,
                frameWidth: Float, animFps: Int, private val frameCount: Int, pixelsPerMetre: Int) {
    var bitmapSheet: Bitmap? = null
    private val bitmapName: String
    private val soursRect: Rect
    private var currentFrame = 0
    private var frameTicker: Long
    private val framePeriod: Int
    private val frameWidth: Float = frameWidth * pixelsPerMetre
    private val frameHeight: Float = frameHeight * pixelsPerMetre
    private val pixelsPerMetre: Int
    fun getCurrentFrame(time: Long, xVelocity: Float,
                        yVelocity: Float, moves: Boolean): Rect {
        if (xVelocity != 0f || yVelocity != 0f || !moves) {
            if (time > frameTicker + framePeriod) {
                frameTicker = time
                currentFrame++
                if (currentFrame >= frameCount) {
                    currentFrame = 0
                }
            }
        }
        soursRect.left = currentFrame * frameWidth.toInt()
        soursRect.right = soursRect.left + frameWidth.toInt()
        return soursRect
    }

    init {
        soursRect = Rect(0, 0, this.frameWidth.toInt(), this.frameHeight.toInt())
        framePeriod = 1000 / animFps
        frameTicker = 0L
        this.bitmapName = "" + bitmapName
        this.pixelsPerMetre = pixelsPerMetre
    }
}